src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1289,51 +1288,42 @@ internal static bool IsGenericTaskType(this TypeSymbol type, CSharpCompilation c
             {
                 return true;
             }
-            NamedTypeSymbol builderType;
-            MethodSymbol createBuilderMethod;
-            return namedType.IsCustomTaskType(out builderType, out createBuilderMethod);
+            object builderArgument;
+            return namedType.IsCustomTaskType(out builderArgument);
         }
 
         /// <summary>
-        /// Returns true if the type is generic or non-generic task-like type. If so, the async
-        /// method builder type is returned along with the method to construct that type.
+        /// Returns true if the type is generic or non-generic custom task-like type due to the
+        /// [AsyncBuilder(typeof(B))] attribute. It returns the "B".
         /// </summary>
-        internal static bool IsCustomTaskType(this NamedTypeSymbol type, out NamedTypeSymbol builderType, out MethodSymbol createBuilderMethod)
+        /// <remarks>
+        /// For the Task types themselves, this method might return true or false depending on mscorlib.
+        /// The definition of "custom task-like type" is one that has an [AsyncBuilder(typeof(B))] attribute,
+        /// no more, no less. Validation of builder type B is left for elsewhere. This method returns B
+        /// without validation of any kind.
+        /// </remarks>
+        internal static bool IsCustomTaskType(this NamedTypeSymbol type, out object builderArgument)
```

Comment:
On the matter of returning object:

We determined that the *parameter* of the attribute is a type. We didn't determine that the *argument* of this particular invocation is a type. Consider this case:
```cs
class AsyncBuilderAttribute : Attribute { public AsyncBuilderAttribute(Type t) {} }
[AsyncBuilder("a")] class Tasklike { }
```
Here the argument is an object, a string literal. I figured it would be a desirable thing for someone to be able to report the error `Argument must be a type, not '%1'`. If the out parameter were a `TypeSymbol` then the caller would have no way of recovering what `%1` should be.

You talked about what the callers do. Ultimately, most of the callers don't care about the argument and they look solely for the attribute. The callers during codegen in `AsyncBuilderCollection` actually require a `NamedTypeSymbol`, not just a `TypeSymbol`. So regardless of whether the builder argument is returned as an `object` or a `TypeSymbol` they'd still be doing an `as` check on it.

The only thing that wouldn't require the caller to do an `as` check is if the builder argument were returned as a `NamedTypeSymbol`. This would still require the codegen callers to do a `null` check anyway, and would prevent them from being able to report errors about what the argument actually was, and wouldn't be needed for the pre-codegen callers.